### ISSUE TITLE
I got every value -1 when I predicted.

### ISSUE THREAD

**Issue by DayChan** [fizyr/keras-retinanet#618]
I used my own dataset to train and I have used the debug.py finding that the data is fine.

But after several times trying, I always got every value -1 when I predicted the result.

Here is the result which my `predict.py` print:

```
Using TensorFlow backend.
2018-08-02 13:30:07.229137: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2018-08-02 13:30:07.327585: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:898] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-08-02 13:30:07.327857: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1344] Found device 0 with properties: 
name: GeForce GTX 1060 6GB major: 6 minor: 1 memoryClockRate(GHz): 1.7845
pciBusID: 0000:01:00.0
totalMemory: 5.93GiB freeMemory: 5.05GiB
2018-08-02 13:30:07.327871: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1423] Adding visible gpu devices: 0
2018-08-02 13:30:07.497702: I tensorflow/core/common_runtime/gpu/gpu_device.cc:911] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-02 13:30:07.497734: I tensorflow/core/common_runtime/gpu/gpu_device.cc:917]      0 
2018-08-02 13:30:07.497741: I tensorflow/core/common_runtime/gpu/gpu_device.cc:930] 0:   N 
2018-08-02 13:30:07.497911: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 4815 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1060 6GB, pci bus id: 0000:01:00.0, compute capability: 6.1)
/usr/local/lib/python3.5/dist-packages/keras/engine/saving.py:269: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
  warnings.warn('No training configuration found in save file: '
scale: 0.15625
processing time:  2.0690972805023193
box: [-6.4 -6.4 -6.4 -6.4]
score: -1.0
label: -1
box: [-6.4 -6.4 -6.4 -6.4]
score: -1.0
label: -1
box: [-6.4 -6.4 -6.4 -6.4]
score: -1.0
label: -1
box: [-6.4 -6.4 -6.4 -6.4]
score: -1.0
label: -1
box: [-6.4 -6.4 -6.4 -6.4]
score: -1.0
label: -1
box: [-6.4 -6.4 -6.4 -6.4]
score: -1.0
label: -1
......(repeat tens of times)
```

My predict code is: `predict.py`
``` python
import keras
import os
import sys
# import keras_retinanet
from keras_retinanet import models
from keras_retinanet.utils.image import read_image_bgr, preprocess_image, resize_image
from keras_retinanet.utils.visualization import draw_box, draw_caption
from keras_retinanet.utils.colors import label_color
from keras.models import Model
# Allow relative imports when being executed as script.
if __name__ == "__main__" and __package__ is None:
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
    import keras_retinanet.bin  # noqa: F401
    __package__ = "keras_retinanet.bin"

from ..models import resnet
import matplotlib.pyplot as plt
import cv2
import os
import numpy as np
import time


import tensorflow as tf


def get_session():
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    return tf.Session(config=config)

keras.backend.tensorflow_backend.set_session(get_session())

model_path = os.path.join('.', 'snapshots','weights' ,'RetinaNet_ver1.hdf5')


model = models.load_model(model_path, backbone_name='resnet152', convert=True)


labels_to_names = {0: 'wrong'}

image = read_image_bgr('./test_ori/J01_2018.06.25 13_29_53.jpg')

draw = image.copy()
draw = cv2.cvtColor(draw, cv2.COLOR_BGR2RGB)

image = preprocess_image(image)
image, scale = resize_image(image,max_side = 400)
print("scale:",scale)

start = time.time()
boxes, scores, labels = model.predict_on_batch(np.expand_dims(image, axis=0))
print("processing time: ", time.time() - start)

boxes /= scale

for box, score, label in zip(boxes[0], scores[0], labels[0]):
    # scores are sorted so we can break
    print("box:",box)
    print("score:",score)
    print("label:",label)
    # if score < 0.5:
    #     continue
        
    color = label_color(label)
    
    b = box.astype(int)
    draw_box(draw, b, color=color)
    
    #caption = "{} {:.3f}".format(labels_to_names[label], score)
    #draw_caption(draw, b, caption)
    
plt.figure(figsize=(15, 15))
plt.axis('off')
plt.imshow(draw)
plt.show()
```
My trianing code is: `train.py`
```python
import argparse
import os
import sys
import warnings

import keras
import keras.preprocessing.image
import tensorflow as tf
from keras.callbacks import ModelCheckpoint, TensorBoard, EarlyStopping
# Allow relative imports when being executed as script.
if __name__ == "__main__" and __package__ is None:
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
    import keras_retinanet.bin  # noqa: F401
    __package__ = "keras_retinanet.bin"

from .. import layers  # noqa: F401
from .. import losses
from .. import models
from ..callbacks import RedirectModel
from ..callbacks.eval import Evaluate
from ..models.retinanet import retinanet_bbox
from ..preprocessing.csv_generator import CSVGenerator
from ..preprocessing.kitti import KittiGenerator
from ..preprocessing.open_images import OpenImagesGenerator
from ..preprocessing.pascal_voc import PascalVocGenerator
from ..utils.anchors import make_shapes_callback
from ..utils.keras_version import check_keras_version
from ..utils.model import freeze as freeze_model
from ..utils.transform import random_transform_generator
from ..models import resnet
from keras.utils import multi_gpu_model


batch_size=4
image_side = 600
train_generator = CSVGenerator(
    'train.csv',
    csv_class_file='classes.csv',
    base_dir='.',
    batch_size=batch_size,
    image_max_side=image_side)
validation_generator = CSVGenerator(
    'validation.csv',
    csv_class_file='classes.csv', 
    base_dir='.',
    batch_size=batch_size,
    image_max_side=image_side)

#epochs=100
ts = 1723 // batch_size
vs = 559 // batch_size
print("ts:",ts)
print(vs)

checkpointer = ModelCheckpoint(filepath = "./snapshots/checkpointer/"+'RetinaNet.{epoch:02d}-{val_loss:.2f}.hdf5',monitor='val_loss', save_best_only=True, save_weights_only=False, period = 1)
#earlyStopping=EarlyStopping(monitor='val_loss', min_delta=0.002, patience=6, verbose=0, mode='auto')
board=TensorBoard(log_dir='./logs')

model = resnet.resnet_retinanet(num_classes = 1 ,backbone = "resnet152")
# model = vgg_retinanet(2)
model = multi_gpu_model(model, gpus=2)


#model.load_weights("./snapshots"+'/weights/RetinaNet_ver0.hdf5')


print("train 0")
model.compile(
    loss={
        'regression'    : losses.smooth_l1(),
        'classification': losses.focal()
    },
    optimizer=keras.optimizers.adam(lr=1e-2, clipnorm=0.001),
    metrics = ["accuracy"]
)
model.fit_generator(train_generator,
                    epochs=50,
                    verbose=1,
                    steps_per_epoch=ts,
                    validation_data=validation_generator,
                    validation_steps = vs,
                    callbacks = [checkpointer,board],
                    initial_epoch = 0)

model.save("./snapshots"+'/weights/RetinaNet_ver0.hdf5')
print('Save successfully0!')

```

My loss and acc is :
```
regression_loss: 2.6712
classification_loss: 1.2559
regression_acc: 0.5280
classification_acc: 0.9845
val_regression_loss: 2.6315
val_classification_loss: 1.1332
val_regression_acc: 0.5028
val_classification_acc: 0.9845
```

I think my loss is not very well, but still it's unusual to get each result -1.
Did I ignore something?

**Comment by hgaiser** [comment]
Thank you for the detailed issue! The result of -1 means that there is no result. I noticed two things, your max image side is different in training and testing, these should ideally be equal. Second, your learning rate seems too high, I'm guessing the model simply doesn't learn anything because of that. Try a learning rate of `1e-5`>

**Comment by DayChan** [comment]
@hgaiser Thank you very much for your answer! 
It's really nice to find my `max_image_side` is wrong, but it still didn't work after I corrected it. 
I will try to check my dataset and train again with the lower learning rate. 
Closing my issue temporarily.

**Comment by hgaiser** [comment]
Alright, sounds good. For the record, your losses are way too high. I
expect the total loss to be a bit below 1.0. That's why I think the real
problem is somewhere in your training.

**Comment by patrickcgray** [comment]
@DayChan I'm curious what (if anything) resolved your issue here. Was it simply more training?

**Comment by DayChan** [comment]
@patrickcgray I think my dataset has not enough data. Simply more training didn't lower my loss at all. So it's nothing changed in the predicting stage. I'm trying to gain more relative data.

**Comment by borakrc** [comment]
Did you run debug.py and check your annotations? They are supposed to be mostly green.

For most cases, 2000 pictures sounds pretty enough to me. How many classes and object instances of each class do you have?

By the way, your if regression accuracy is 0.53 and classification accuracy is 0.98, you should be seeing some results at prediction.

**Comment by DayChan** [comment]
@borakrc I did run the debug.py, and it seems coming out that the frame is red. How can I find out what problem it has?

**Comment by linxuefeng-hust** [comment]
@DayChan  I meet the same problem ,too. And I don't the reason is the loss. 
Because my train loss is 0.044. Did you find the way to solve the problem?

**Comment by DayChan** [comment]
@linxuefeng-hust 
Noticed that you are probably Chinese, I would reply in Chinese, if not, please tell me.
实际上这可能跟训练集的有关，如果训练集的目标占比太小，可能就会有这个问题。你可以用`debug.py`检查一下是不是这个问题，如果目标的框太小的话，这个py程序显示出来的框就是红色的，如果没问题就是绿色的。

**Comment by linxuefeng-hust** [comment]
谢谢你的回复。我做的是工业瑕疵检测，所以目标相对会占比较小。但我用debug.py（ python keras_retinanet/bin/debug.py csv ./csv/annotations.csv ./csv/classes.csv）检查的时候，确实出现了一张红色的框图片，但也只出现了一张图片，也没有出现绿色框的图片，我不确定是不是debug.py没有运行成功。根据readme上写的，少量的红框应该不影响训练，所以我不太确定是不是数据的问题？

**Comment by DayChan** [comment]
> 谢谢你的回复。我做的是工业瑕疵检测，所以目标相对会占比较小。但我用debug.py（ python keras_retinanet/bin/debug.py csv ./csv/annotations.csv ./csv/classes.csv）检查的时候，确实出现了一张红色的框图片，但也只出现了一张图片，也没有出现绿色框的图片，我不确定是不是debug.py没有运行成功。根据readme上写的，少量的红框应该不影响训练，所以我不太确定是不是数据的问题？

这就不是很清楚了，我这个issue也是做的一个瑕疵检测的训练然后预测不出来，后来也没解决。只是后来我用这个来训练一些稍微大些的目标就没有出现这个问题了。

**Comment by hgaiser** [comment]
Please keep to using English if you want other people to join the conversation. If you don't want other people to join the conversation, I suggest using Slack and message each other directly.

**Comment by linxuefeng-hust** [comment]
> > 谢谢你的回复。我做的是工业瑕疵检测，所以目标相对会占比较小。但我用debug.py（ python keras_retinanet/bin/debug.py csv ./csv/annotations.csv ./csv/classes.csv）检查的时候，确实出现了一张红色的框图片，但也只出现了一张图片，也没有出现绿色框的图片，我不确定是不是debug.py没有运行成功。根据readme上写的，少量的红框应该不影响训练，所以我不太确定是不是数据的问题？
> 
> 这就不是很清楚了，我这个issue也是做的一个瑕疵检测的训练然后预测不出来，后来也没解决。只是后来我用这个来训练一些稍微大些的目标就没有出现这个问题了。

多谢解答了。原来是这个project本身存在的问题啊。。。当时就是想到retinanet在小目标检测上表现比yolo好才使用的这个模型，没想到坑在这里面了。

**Comment by linxuefeng-hust** [comment]
> Please keep to using English if you want other people to join the conversation. If you don't want other people to join the conversation, I suggest using Slack and message each other directly.

Both  of  us have meet the same problem when using this project to detect the small size things. It seems like that the model trained by the same size objections can't predict well. And we don't find the reason.

**Comment by DayChan** [comment]
> > > 谢谢你的回复。我做的是工业瑕疵检测，所以目标相对会占比较小。但我用debug.py（ python keras_retinanet/bin/debug.py csv ./csv/annotations.csv ./csv/classes.csv）检查的时候，确实出现了一张红色的框图片，但也只出现了一张图片，也没有出现绿色框的图片，我不确定是不是debug.py没有运行成功。根据readme上写的，少量的红框应该不影响训练，所以我不太确定是不是数据的问题？
> > 
> > 
> > 这就不是很清楚了，我这个issue也是做的一个瑕疵检测的训练然后预测不出来，后来也没解决。只是后来我用这个来训练一些稍微大些的目标就没有出现这个问题了。
> 
> 多谢解答了。原来是这个project本身存在的问题啊。。。当时就是想到retinanet在小目标检测上表现比yolo好才使用的这个模型，没想到坑在这里面了。

It seems that there is another possibility that the object is similar with the background which make it hard to be recognized.

**Comment by hgaiser** [comment]
With defects that could very well be the case.

**Comment by linxuefeng-hust** [comment]
@DayChan  But , which confused me is the train loss is small. If the object can't be recognized , the train loss should be bigger? Did you use the transfer learning by coco_weights?
